### PR TITLE
watchers: don't dereference symlinks when copying files

### PR DESCRIPTION
**Pull request by egernst**
The current implementation just copies the file, dereferencing any
  simlinks in the process. This results in symlinks no being preserved,
  and a change in layout relative to the mount that we are making
  watchable.
    
 What we want is something like "cp -d"
    
This isn't available in a crate, so let's go ahead and introduce a copy
function which will create a symlink with same relative path if the
source file is a symlink. Regular files are handled with the standard
fs::copy.
   
 Introduce a unit test to verify symlinks are now handled 

Fixes: #2950

### CODE REVIEW

File: src/agent/src/watcher.rs [hunk 1]
```diff
@@ -110,21 +120,7 @@ impl Storage {
...
-        fs::copy(&source_file_path, &dest_file_path)
-            .await
-            .with_context(|| {
-                format!(
-                    "Copy from {} to {} failed",
-                    source_file_path.display(),
-                    dest_file_path.display()
-                )
-            })?;
+        copy(&source_file_path, &dest_file_path).await?;
```
Comment (sameo) [inline src/agent/src/watcher.rs:123]: Could we keep the error message?

File: src/agent/src/watcher.rs [hunk 2]
```diff
@@ -843,6 +839,95 @@ mod tests {
...
+        tokio::fs::symlink(&src_file, &src_symlink_file)
+            .await
+            .unwrap();
+        copy(src_symlink_file, &dst_symlink_file).await.unwrap();
+        // verify destination:
+        assert!(fs::symlink_metadata(&dst_symlink_file)
+            .unwrap()
+            .file_type()
+            .is_symlink());
+        assert_eq!(fs::read_link(dst_symlink_file).unwrap(), src_file);
```
Comment (sameo) [inline src/agent/src/watcher.rs:871]: You want to take a reference on `dst_symlink_file` here and not move it.

**Review by sameo — APPROVED** [review]

**Review by mxpv — APPROVED** [review]

**Comment by egernst** [comment]
We should also make sure we treat directory symlinks similar to a file. This'll require a bit more refactoring.

Moved back to draft.

**Comment by fidencio** [comment]
```
thread 'watcher::tests::watch_directory_with_symlinks' panicked at 'assertion failed: `(left == right)`
  left: `0`,
 right: `1`', src/watcher.rs:1005:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test watcher::tests::watch_directory_with_symlinks ... FAILED
test watcher::tests::watch_file ... ok
test watcher::tests::watch_directory_too_large ... ok
test util::tests::test_interruptable_io_copier_reader ... ok
test watcher::tests::verify_container_cleanup_watching ... ok

failures:

failures:
    watcher::tests::watch_directory_with_symlinks

test result: FAILED. 92 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out

error: test failed, to rerun pass '-p kata-agent --bin kata-agent'
```
@Jakob-Naucke already mentioned that such failures may be happening due to the tests running in parallel on BM.

**Comment by Jakob-Naucke** [comment]
yeeeeah it's possible this PR introduces a sibling to #2946, which I'm not super happy about. We'll have to fix it in the same way I guess.

**Comment by fidencio** [comment]
Same issue.  @egernst, mind to hold this for one day?  It will give time to @Jakob-Naucke either debug it, or let you know that this test would be good candidate to be skipped on s390x.  By doing this we avoid breaking s390x CI entirely.

**Comment by egernst** [comment]
debugging s390x failure:
```
thread 'watcher::tests::watch_directory_with_symlinks' panicked at 'assertion failed: `(left == right)`
  left: `0`,
 right: `1`', src/watcher.rs:1005:9
 ```
 
 This seems same condition brought up in https://github.com/kata-containers/kata-containers/issues/2946

**Pull request merged**